bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java

```diff
@@ -156,4 +143,23 @@ public static void checkIsMapOnlyOperation(Type mapType, String op) {
                 throw createOpNotSupportedError(mapType, op);
         }
     }
+
+    private static void updateMapValue(Type mapType, MapValue<BString, Object> mapValue, BString fieldName,
+                                       Object value) {
+        switch (mapType.getTag()) {
+            case TypeTags.MAP_TAG:
+                handleInherentTypeViolatingMapUpdate(value, (BMapType) mapType);
+                mapValue.put(fieldName, value);
+                return;
+            case TypeTags.RECORD_TYPE_TAG:
+                if (handleInherentTypeViolatingRecordUpdate(mapValue, fieldName, value, (BRecordType) mapType, false)) {
+                    mapValue.put(fieldName, value);
+                    return;
+                }
+                mapValue.remove(fieldName);
+                return;
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG:
+                updateMapValue(((BTypeReferenceType) mapType).getReferredType(), mapValue, fieldName, value);
```

Comment:
Shall we add another test where the referred type is another type reference type?

```ballerina
public type StringMap map<string>;
public type Map StringMap;

Map x = {};

function testFn()  {
    x["greeting"] = "hello";
    // assertion
}
```

Reply:
This scenario covered in the example that was added. Right?

Reply:
Ah yes, my bad. Shall we add one for just Map though?

```ballerina
function testFn()  {
    Map x = {};
    x["greeting"] = "hello";
    // assertion
}
```